montecarlo/vmc/src/TGeoMCBranchArrayContainer.cxx

```diff
@@ -0,0 +1,169 @@
+// @(#)root/vmc:$Id$
+// Authors: Benedikt Volkel 07/03/2019
+
+/*************************************************************************
+ * Copyright (C) 2006, Rene Brun and Fons Rademakers.                    *
+ * Copyright (C) 2002, ALICE Experiment at CERN.                         *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+/** \class TGeoMCBranchArrayContainer
+    \ingroup vmc
+
+Storing and re-using geometry states of the TGeoManager in use by storing
+them as TGeoBranchArrays.
+
+After having initialized a navigator using a stored state, it can be freed
+to be used again for storing another geometry state. This makes it easy to
+handle many events with many stored geometry states and the memory used is
+kept as small as possible.
+*/
+
+#include "TGeoMCBranchArrayContainer.h"
+#include "TGeoManager.h"
+#include "TGeoBranchArray.h"
+#include "TError.h"
+
+ClassImp(TGeoMCBranchArrayContainer);
+
+TGeoMCBranchArrayContainer::TGeoMCBranchArrayContainer()
+   : fCacheSize(0), fInitCacheSize(0), fNextIndex(0), fCurrentIndex(-1), fMaxLevels(0), fIsInitialized(kFALSE)
+{
+}
+
+TGeoMCBranchArrayContainer::~TGeoMCBranchArrayContainer()
+{
+   for (auto &c : fCache) {
+      if (c) {
+         delete c;
+      }
+   }
+}
+
+void TGeoMCBranchArrayContainer::Initialize(Int_t maxLevels, Int_t size)
+{
+   fCacheSize = size;
+   fMaxLevels = maxLevels;
+   if (fIsInitialized) {
+      ResetCache();
+   }
+   // All indices are free at this point.
+   fIsIndexFree.resize(fCacheSize, kTRUE);
```

Comment:
True, I can get rid of some state.
However, I would like to keep TGeoBranchArray objects once they have been created since they can be re-used. In this way, I don't free and allocate memory all the time I store and use these objects to initialize a TGeoNavigator.
Also, I don't want to look for the first null entry, because geometry states may be stored for many tracks. This could mean that such a loop would be called 1000s of times having >100 entries in the vector. That would scale roughly with N_events*N_tracks*N_changingBetweenEngines

Reply:
State monitoring has been made more concise